### PR TITLE
Update RARE reconstruction algorithm to allow free variables defined in conditions

### PR DESCRIPTION
**Pull request by ajreynol**
Allows RARE rules to contain free variables x not in the left hand side of a rule, provided they occur defined in a condition e.g. an equality x = t, where t contains only variables that are in the left hand side of the rule.

This updates the README for RARE.

It also fixes an inconsistency between the evaluator and the rewriter for `ints.log2` which was discovered while testing this change.

### CODE REVIEW

File: src/theory/arith/arith_rewriter.cpp [hunk 1]
```diff
@@ -968,7 +968,8 @@ RewriteResponse ArithRewriter::postRewriteIntsLog2(TNode t)
     Assert(t[0].getType().isInteger());
     Integer i = t[0].getConst<Rational>().getNumerator();
     size_t const length = i.length();
-    return RewriteResponse(REWRITE_DONE, rewriter::mkConst(Integer(length)));
+    return RewriteResponse(REWRITE_DONE,
+                           rewriter::mkConst(Integer(length - 1)));
```
Comment (abdoo8080) [inline src/theory/arith/arith_rewriter.cpp:972]: Interesting bug. Maybe add an assertion that `i` is positive? (unless it's defined for non-positive numbers, in which case I would prefer the result to be 0)

Comment (ajreynol) [reply src/theory/arith/arith_rewriter.cpp:1014]: I've update to make log of negative equal to 0.

Comment (bdalrhm) [reply src/theory/arith/arith_rewriter.cpp:1014]: What about 0?

Comment (ajreynol) [reply src/theory/arith/arith_rewriter.cpp:1014]: log_2 0 was zero already.

**Review by abdoo8080 — APPROVED** [review]

**Pull request merged**